lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,131 @@
+-}
+getSubs :: Pat -> Pat -> [(Text, Pat)] -- should be a set, whatever
+getSubs (VarP s) t = [(s, t)]
+getSubs (x :* xs) (x' :* xs') = getSubs x x' <> getSubs xs xs'
+getSubs (l := t) (l' := t') = getSubs l l' <> getSubs t t'
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+-- NoMatch isn't fatal but OverlappingMatches is (i.e. we need to stop when we encounter it)
+data MatchError
+  = NoMatch
+  | OverlappingMatches [Rule]
+
+-- for SolveM, since we catch NoMatch
+data Overlap = Overlap Constraint [Rule]
+  deriving stock (Show, Eq)
+
+selectMatchingInstance :: Pat -> Class -> [Rule] -> Either MatchError Rule
+selectMatchingInstance p c rs = case filter matchPatAndClass rs of
+  [] -> Left NoMatch
+  [r] -> Right r
+  overlaps -> Left $ OverlappingMatches overlaps
+  where
+    matchPatAndClass :: Rule -> Bool
+    matchPatAndClass r =
+      ruleHeadClass r == c
+        && ruleHeadPat r
+        `matches` p
+
+type SolveM = ReaderT [Rule] (WriterT (S.Set Constraint) (Either Overlap))
+
+{- Given a list of instances (the initial scope), determines whether we can derive
+   an instance of the Class argument for the Pat argument. A result of [] indicates that there are
+   no remaining subgoals and that the constraint has been solved.
+   NOTE: At the moment this handles superclasses differently than you might expect -
+         instead of assuming that the superclasses for all in-scope classes are defined,
+         we check that those constraints can be solved before affirmatively judging that the
+         target constraint has been solved. I *think* that makes sense in this context (whereas in Haskell
+         it doesn't b/c it's *impossible* to have `instance Foo X` if the definition of Foo is
+         `class Bar y => Foo y` without an `instance Bar X`)
+-}
+solveM :: Constraint -> SolveM ()
+solveM cst@(C c pat) =
+  ask >>= \inScope ->
+    -- First, we look for the most specific instance...
+    case selectMatchingInstance pat c inScope of
+      Left e -> case e of
+        NoMatch -> tell $ S.singleton cst
+        OverlappingMatches olps -> throwError $ Overlap cst olps
+      -- If there is, we substitute the argument of the constraint to be solved into the matching rules
+      Right rule -> case subst rule pat of
+        -- If there are no additional constraints on the rule, we try to solve the superclasses
+        C _ p :<= [] -> solveClassesFor p (csupers c)
+        -- If there are additional constraints on the rule, we try to solve them
+        C _ _ :<= is -> do
+          traverse_ solveM is
+          solveClassesFor pat (csupers c)
```

Comment:
This reads nicer thanks.

I would strongly state that the `super class` constraints should be part of the rules themselves and not treat them specially. Specifically how I was doing it

```
ord(maybe(A)) :-
  ord(A),
  eq(maybe(A)).
```

And then the machinery just works as intended, no need to consider it specially right?

Reply:
That way you solver doesn't need any knowledge of what a superclass is, it's just a simple rule system.

Reply:
Well, maybe. I missed something here: For all of the (other) instances that we assume to hold when checking a particular instance, it's probably safe to assume that their superclasses hold as well. I'm not 100% sure on this though and it needs more thought, but it's not really a problem *here* so much as it is a problem for #44